clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1398,6 +1409,100 @@ event_delay_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t
     return DR_EMIT_DEFAULT;
 }
 
+
+typedef struct _func_metadata {
+    const trace_marker_type_t marker_retaddr;
+    const trace_marker_type_t marker_arg;
+    const trace_marker_type_t marker_retval;
+    const char *name;
+    const int num_args;
+} func_metadata;
+
+#define MALLOC_RETADDR TRACE_MARKER_TYPE_FUNC_MALLOC_RETADDR
+#define MALLOC_ARG     TRACE_MARKER_TYPE_FUNC_MALLOC_ARG
+#define MALLOC_RETVAL  TRACE_MARKER_TYPE_FUNC_MALLOC_RETVAL
+
+static const func_metadata instru_funcs[] = {
+    {MALLOC_RETADDR, MALLOC_ARG, MALLOC_RETVAL, "malloc", 1},
+    {MALLOC_RETADDR, MALLOC_ARG, MALLOC_RETVAL, "(anonymous namespace)::do_malloc", 1},
+
+#ifdef WINDOWS
+    {MALLOC_RETADDR, MALLOC_ARG, MALLOC_RETVAL, "malloc_impl", 1 },
+#endif // WINDOWS
+
+#ifdef UNIX
+    {MALLOC_RETADDR, MALLOC_ARG, MALLOC_RETVAL, "tc_malloc", 1 },
+    {MALLOC_RETADDR, MALLOC_ARG, MALLOC_RETVAL, "__libc_malloc", 1},
```

Comment:
I think we want to be able to dynamically specify which functions to record: perhaps via a config file eventually, though we could start with a runtime option.  We just need the name and number of args if all we support is recording pointer-sized arg values.  You could imagine sthg like:

```
-record_function malloc 1 -record_function realloc 2
```

If we decide heap allocation is very common and we want to have built-in support, I would say it should be under a separate option like `-record_function_malloc` or `-record_heap_allocs` or sthg.

Alternatively, we could expose an interface to write metadata into the trace and structure all of this as a separate client from the memory address tracer.  I'm fine bundling it in for now as it seems like a common thing to want to augment an address trace with.

Reply:
Done